esx_service/utils/vm_listener.py

```diff
@@ -0,0 +1,205 @@
+#!/usr/bin/env python
+# Copyright 2016 VMware, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+'''
+VM change listener (started as a part of vmdkops service).
+It monitors VM poweroff events and detaches the DVS managed
+volumes from the VM and updates the status in KV
+'''
+
+import logging
+import os
+import os.path
+import atexit
+import time
+import http.client
+
+import threadutils
+import log_config
+import vmdk_utils
+import vmdk_ops
+
+from pyVmomi import VmomiSupport, vim, vmodl
+# vim api version used - version11
+
+VM_POWERSTATE = 'runtime.powerState'
+POWERSTATE_POWEROFF = 'poweredOff'
+HOSTD_RECONNECT_INTERVAL = 5 #approx time for hostd to comeup is 10-15 seconds
+HOSTD_RECONNECT_ATTEMPT = 5
+
+
+def get_propertycollector():
+    """
+    Connect to hostd. If failed, retry.
+    Create the property collector with filter to monitor VM power state changes
+    Return the property collecter and error (if any)
+    """
+
+    si = vmdk_ops.get_si()
+
+    reconnect_interval = HOSTD_RECONNECT_INTERVAL
+    for i in range(HOSTD_RECONNECT_ATTEMPT):
+        if si:
+            break
+
+        # If hostd is not up yet, sleep for a while and try again
+        logging.warn("VMChangeListener couldn't connect to hostd.")
+        logging.warn("Retrying after %s seconds", reconnect_interval)
+        time.sleep(reconnect_interval)
+        si = vmdk_ops.get_si()
+
+        # exponential backoff for next retry
+        reconnect_interval += reconnect_interval
+
+    # Proceed further only after you get si instance
+    if not si:
+        # could not connect to hostd even after retries
+        # Something is seriously wrong
+        return None, "Hostd Down!"
```

Comment:
"Hostd Down!" => "Unable to connect to hostd. Verify that vmware-hostd is running."